### PR TITLE
DDP-5574 update create instance endpoint to require parent instance

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

See DDP-5574. This PR updates the [CreateActivityInstance](https://pepper-dev.datadonationplatform.org/docs#operation/createActivity) API endpoint to check for parent instance when creating child instance.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [x] :relaxed: All good, business as usual!

## How do we demo these changes?

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!

### CODE REVIEW

File: pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml [hunk 1]
```diff
@@ -31,7 +31,16 @@ post:
   tags:
     - Activities
   summary: create a new instance of an activity
-  description: Given an `activityCode`, the server will create a new instance of the activity for the specified user if the configured conditions for the activity is met.
```
Comment (yufengwng) [inline pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml:34]: @alextanlan @bskinner I've updated documentation descriptions to better describe parent/child dependency when creating child instances, particularly about read-only behavior.

Comment (bskinner) [reply pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml:34]: Thank you for the heads up

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java [hunk 2]
```diff
@@ -52,22 +64,54 @@ public Object handle(Request request, Response response, ActivityInstanceCreatio
...
+            }
+
+            Long parentInstanceId = null;
+            if (validation.getParentActivityCode() != null) {
+                if (StringUtils.isBlank(parentInstanceGuid)) {
+                    String msg = "Creating child nested activity instance requires parent instance guid";
+                    LOG.warn(msg);
+                    throw ResponseUtil.haltError(response, HttpStatus.SC_BAD_REQUEST,
+                            new ApiError(ErrorCodes.BAD_PAYLOAD, msg));
+                }
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java]: This logging/exception throwing lines could be extracted to method (repeated 3 times in this class):
```
private void haltError(int status, String code, String msg) {
        LOG.warn(msg);
        throw ResponseUtil.haltError(status, new ApiError(code, msg));
}       
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java [hunk 3]
```diff
@@ -52,22 +64,54 @@ public Object handle(Request request, Response response, ActivityInstanceCreatio
                 var err = new ApiError(ErrorCodes.ACTIVITY_NOT_FOUND,
                         "Could not find creation validation information for activity " + activityCode);
                 LOG.warn(err.getMessage());
-                throw ResponseUtil.haltError(response, 404, err);
+                throw ResponseUtil.haltError(response, HttpStatus.SC_NOT_FOUND, err);
+            }
+
+            Long parentInstanceId = null;
+            if (validation.getParentActivityCode() != null) {
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java:73]: I'd suggest to extract code under this `if` to method `validateParentActivity(...)`.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceDao.sql.stg [hunk 4]
```diff
@@ -2,6 +2,11 @@ group ActivityInstanceDao;
 
 queryActivityInstanceCreationValidation() ::= <<
 select act.study_activity_id,
+       act.parent_activity_id,
+       (select study_activity_code
+          from study_activity
+         where study_activity_id = act.parent_activity_id
+       ) as parent_activity_code,
```
Comment (evgeniipr) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceDao.sql.stg:9]: Not sure about performance (needs to compare execution plans) but seems it's possible to replace nested select with
left join study_activity sa on sa.study_activity_id = act.parent_activity_id.
In this post there are opinions that `left join` is better by performance that `subquery`: https://stackoverflow.com/questions/2577174/join-vs-sub-query
For example, one guy said:
`As per my observation like two cases, if a table has less then 100,000 records then the join will work fast.
But in the case that a table has more than 100,000 records then a subquery is best result.`

**Review by MocanaAtBroad — DISMISSED** [review]

**Review by evgeniipr — DISMISSED** [review]

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceDao.sql.stg (hunk 4 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceDao.sql.stg:9]: Most activities will not have a parent activity, so most of the time the subquery will be querying with `null` as `act.parent_activity_id`. I'm hoping the DB can see that it's null and optimize out the subquery. I took a brief look of the query plan as I was writing this query, and I didn't see any immediate red flags, so I think I'll leave this query as-is for now.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java]: All set! 6238d092369add6c34bf28cd07d5aff6ce0d9be3

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java:73]: Done! 6238d092369add6c34bf28cd07d5aff6ce0d9be3

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**